File: 53-MaxSubarray/v2.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int n = nums.size();
        int currSum = nums[0], maxSum = nums[0];

        for(int i = 1; i < n; i++) {
            currSum = max(nums[i], currSum + nums[i]);
            maxSum = max(maxSum, currSum);
        }
        return maxSum;
    }
};

int main()
{
    Solution solu;
    vector<int> nums = {6,7,-15,7,16};
    cout << solu.maxSubArray(nums) << endl;
    return 0;
}

File: 55-JumpGame/v2.java
class Solution {
    public boolean canJump(int[] nums) {
        int len = nums.length;
        int curEnd = 0, curFurthest = 0; //定位器，Furthest表示可达到的最远距离，End表示目前的目标地点
        for (int i = 0; i < len; i++) {
            curFurthest = Math.max(curFurthest, i + nums[i]);
            if (i == curEnd) {
                curEnd = curFurthest;
                if(i == curFurthest) break; //不能继续往下走了
            }
        }
        return curEnd >= len - 1 ? true : false;
    }

    public static void main(final String[] args) {
        final var nums = new int[] { 2, 5, 0, 0 };
        final var solu = new Solution();
        System.out.println(solu.canJump(nums));
    }
}